src/Cardano/Wallet/CoinSelection/LargestFirst.hs

```diff
@@ -0,0 +1,109 @@
+largestFirst opt utxo txOutputs = do
+    let txOutputsSorted = NE.toList $ NE.sortBy (flip $ comparing coin) txOutputs
+    let n = fromIntegral $ maximumNumberOfInputs opt
+    let nLargest = take n . L.sortBy (flip $ comparing (coin . snd)) . Map.toList . getUTxO
+    let moneyRequested = sum $ (getCoin . coin) <$> txOutputsSorted
+    let utxoBalance = fromIntegral $ balance utxo
+    let numberOfUtxoEntries = fromIntegral $ L.length $ (Map.toList . getUTxO) utxo
+    let numberOfTransactionOutputs = fromIntegral $ NE.length txOutputs
+
+    when (utxoBalance < moneyRequested)
+        $ throwE $ NotEnoughMoney utxoBalance moneyRequested
+
+    when (numberOfUtxoEntries < numberOfTransactionOutputs)
+        $ throwE $ UtxoNotEnoughFragmented numberOfUtxoEntries numberOfTransactionOutputs
+
+    case foldM atLeast (nLargest utxo, mempty) txOutputsSorted of
+        Just (_, s) ->
+            return s
+        Nothing ->
+            throwE $ MaximumInputsReached (fromIntegral n)
+
+
+-- Selecting coins to cover at least the specified value
+-- The details of the algorithm are following:
+-- (a) transaction outputs are processed starting from the largest one
+-- (b) `maximumNumberOfInputs` biggest available UTxO inputs are taken
+--      into consideration. They constitute a candidate UTxO inputs from
+--      which coin selection will be tried
+-- (c) the biggest candidate UTxO input is tried first to cover the transaction
+--     output. If the input is not enough, then the next biggest one is added
+--     to check if they can cover the transaction output. This process id continued
+--     until the output is covered or the candidates UTxO inputs are depleted.
+--     In the latter case `MaximumInputsReached` error is triggered. If the transaction
+--     output is covered the next biggest one is processed. Here, the biggest
+--     UTxO input, not participating in the coverage, is taken. We are back at (b)
+--     step as a result
+--
+-- The steps are continued until all transaction are covered.
+atLeast
+    :: ([(TxIn, TxOut)], CoinSelection)
+    -> TxOut
+    -> Maybe ([(TxIn, TxOut)], CoinSelection)
+atLeast (utxo0, selection) txout =
+    coverOutput (fromIntegral $ getCoin $ coin txout, mempty) utxo0
+    where
+    coverOutput
+        :: (Int, [(TxIn, TxOut)])
+        -> [(TxIn, TxOut)]
+        -> Maybe ([(TxIn, TxOut)], CoinSelection)
+    coverOutput (target, ins) utxo
+        | target <= 0 = Just
+            ( utxo
+            , selection <> CoinSelection
+                { inputs = ins
+                , outputs = [txout]
+                , change = [Coin (fromIntegral $ abs target)]
+                }
+            )
+        | null utxo =
+            Nothing
+        | otherwise =
+            let
+                (inp, out):utxo' = utxo
+                target' =
+                    target - (fromIntegral (getCoin (coin out)))
+            in
+                coverOutput (target', ins ++ [(inp, out)]) utxo'
```

Comment:
In general, I'd suggest to always prepend elements to lists when using recursion with a list accumulator, and in the terminal case, reverse the list. Because built-in lists in Haskell are linked-list, adding elements to the end of a list is costly, while prepending elements is `O(1)`. Here, it probably doesn't matter much because inputs will remain rather small in practice; but still :no_mouth:

Reply:
done, and also updated tests for it